pkg/syncer/storagepool/storagepoolwatch.go

```diff
@@ -0,0 +1,127 @@
+}
+
+// StartStoragePoolWatch starts a go-routine which processes watch events
+func (w *SPWatcher) StartStoragePoolWatch(ctx context.Context, scWatchCntlr *StorageClassWatch, spCtl *SpController) error {
+	err := w.renewStoragePoolWatch(ctx)
+	if err != nil {
+		return err
+	}
+	go w.watchStoragePool(ctx, scWatchCntlr, spCtl)
+	return nil
+}
+
+// As our watch can and will expire, we need a helper to renew it
+// Note that after we re-new it, we will get a bunch of ADDED events, which may
+// trigger a full remediation
+func (w *SPWatcher) renewStoragePoolWatch(ctx context.Context) error {
+	var err error
+	spClient, spResource, err := getSPClient(ctx)
+	if err != nil {
+		return err
+	}
+	// This means every 24h our watch may expire and require to be re-created.
+	// When that happens, we may need to do a full remediation, hence we change
+	// from 30m (default) to 24h.
+	timeout := int64(60 * 60 * 24) // 24h
+	w.spWatch, err = spClient.Resource(*spResource).Watch(ctx, metav1.ListOptions{
+		TimeoutSeconds: &timeout,
+	})
+	return err
+}
+
+// Handler for storage class watch firing. Handles shutdown and
+// watch renewal, and if it detects any external changes made to StoragePool
+// CRD instance it triggers a full remediation via ReconcileAllStoragePools.
+// external changes are those write operation which are not performed by SPController
+// but by some other entity for example k8s admin user.
+func (w *SPWatcher) watchStoragePool(ctx context.Context, scWatchCntlr *StorageClassWatch, spCtl *SpController) {
+	log := logger.GetLogger(ctx)
+
+	done := false
+	for !done {
+		select {
+		case <-ctx.Done():
+			log.Info("watchStoragePool shutdown", "ctxErr", ctx.Err())
+			done = true
+		case e, ok := <-w.spWatch.ResultChan():
+			if !ok {
+				log.Info("StoragePool watch not ok")
+				err := w.renewStoragePoolWatch(ctx)
+				if err != nil {
+					// XXX: Not sure how to handle this, as we need this watch.
+					// So crash and let restart perform any remediation required
+					log.Error(err, "Fatal event, couldn't renew StoragePool watch")
+					os.Exit(4)
+					return
+				}
+				continue
+			}
+			// run this task on a separate log context that has a separate TraceId for every invocation
+			taskCtx := logger.NewContextWithLogger(ctx)
+			if w.isExternalChange(ctx, e) {
+				log.Info("External change to StoragePool instance detected. Reconciling all StoragePools.")
+				err := ReconcileAllStoragePools(taskCtx, scWatchCntlr, spCtl)
+				if err != nil {
+					log.Errorf("Failed to correct external changes to StoragePool instance. err: %v", err)
+				}
+			}
+		}
+	}
+	log.Info("watchStoragePool ends")
+}
+
+func (w *SPWatcher) isExternalChange(ctx context.Context, e watch.Event) bool {
+	log := logger.GetLogger(ctx)
+	sp, ok := e.Object.(*unstructured.Unstructured)
+	if !ok {
+		log.Debugf("Object in StoragePool watch event is not of type *unstructured.Unstructured, but of type %T", e.Object)
+		return false
+	}
+	spName := sp.GetName()
```

Comment:
Also skip if this StoragePool.spec.driver is not `csi.vsphere.vmware.com` (ie. `csitypes.Name`)

Reply:
done :)